### PR TITLE
Feature posts

### PR DESCRIPTION
**Pull request by jiwan-gharti**
refactored code

### CODE REVIEW

File: posts/models.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    class Meta:
+        abstract = True
+        ordering = ["-created_at"]
+
+
+class Post(TimeStampBaseModel):
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.DO_NOTHING,
+        related_name="post_user",
```
Comment (qmuz9ave) [inline posts/models.py:20]: do changes

File: posts/models.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+from django.db import models
+from django.conf import settings
+
+from ckeditor_uploader.fields import RichTextUploadingField
+
+
+class TimeStampBaseModel(models.Model):
```
Comment (qmuz9ave) [inline posts/models.py:7]: quick-flow ->models.py

File: posts/fields.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
+from django.utils.safestring import mark_safe
+
+class PostDescriptionField(serializers.Field):
```
Comment (qmuz9ave) [inline posts/fields.py:4]: CKeditorSerializerField()

File: posts/permissions.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        if request.method in permissions.SAFE_METHODS:
+            return True
+        return request.user == obj.user
+
+class IsUpvotedByUser(BasePermission):
+    def has_object_permission(self, request, view, obj):
+        if request.method in permissions.SAFE_METHODS:
+            return True
+        elif request.method == "patch":
+            print("patch")
```
Comment (qmuz9ave) [inline posts/permissions.py:17]: remove print

File: posts/serializers.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        model = Post
+        fields = [
+            'title',
+            'description',
+            'tag'
+        ]
+    
+
+class AnswerSerializer(serializers.ModelSerializer):
+    # post = AnswerRelatedPostSerializer(read_only = True)
```
Comment (qmuz9ave) [inline posts/serializers.py:80]: remove comments

File: posts/urls.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+from rest_framework.routers import DefaultRouter
+
+from . import views
+
+app_name = 'posts'
+
+router = DefaultRouter()
+
+router.register('posts',viewset=views.PostView, basename="posts")
```
Comment (qmuz9ave) [inline posts/urls.py:11]: move the routers to root router

File: posts/views.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    authentication_classes = [SessionAuthentication, BasicAuthentication]
+    permission_classes = [IsAuthenticatedOrReadOnly, IsAuthorOrReadOnly, ]
+
+    pagination_class = CustomeLimitOffsetPagination
+    filterset_fields = ['title', 'description']
+
+    
+    def create(self, request, *args, **kwargs):
+        """
+        serializer=aaserializer(data,context={"user":request.user})
```
Comment (qmuz9ave) [inline posts/views.py:47]: remove comment

File: posts/views.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        self.context
+        """
+        data = request.data
+        post = Post.objects.create(
+            user = request.user,
+            title = data['title'],
+            description = data['description'],
+        )
+        post.save()
```
Comment (qmuz9ave) [inline posts/views.py:56]: remove save object

File: posts/views.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            print("Exception: ",e)
+    
+    @action(detail=False, methods = ['patch',],permission_classes = [IsUpvotedByUser,])
+    def upvote_count(self, request, pk = None):
+        try:
+            data = request.data
+            post = Post.objects.filter(id = data['post']).first()
+            if data['upvote_count_signal'] == "up":
+                post.upvote_count = int(post.upvote_count) + 1
+                print(post.upvote_count )
```
Comment (qmuz9ave) [inline posts/views.py:87]: remove pint

File: posts/views.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    serializer_class = PostSerializer
+
+    authentication_classes = [SessionAuthentication, BasicAuthentication]
+    permission_classes = [IsAuthenticatedOrReadOnly, IsAuthorOrReadOnly, ]
+
+    pagination_class = CustomeLimitOffsetPagination
+    filterset_fields = ['title', 'description']
+
+    
+    def create(self, request, *args, **kwargs):
```
Comment (qmuz9ave) [inline posts/views.py:45]: make function shorter and review this function again

File: posts/views.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        
+        except Exception as e:
+            print("Exception: ",e)
+    
+    @action(detail=False, methods = ['patch',],permission_classes = [IsUpvotedByUser,])
+    def upvote_count(self, request, pk = None):
+        try:
+            data = request.data
+            post = Post.objects.filter(id = data['post']).first()
+            if data['upvote_count_signal'] == "up":
```
Comment (qmuz9ave) [inline posts/views.py:85]: data.get("dsds")

File: posts/views.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+                post.is_upvoted = True
+                post.save()
+            serializer = PostSerializer(post)
+            return Response(serializer.data, status=status.HTTP_200_OK)
+        except:
+            return Response({"response":"Bad request"}, status=status.HTTP_400_BAD_REQUEST)
+                    
+        
+
+class AnswerView(viewsets.ModelViewSet):
```
Comment (qmuz9ave) [inline posts/views.py:101]: api/post/{post_id}/answer/

File: posts/views.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    queryset = Answer.objects.all()
+    serializer_class = AnswerSerializer
+    permission_classes = [IsAuthenticatedOrReadOnly, IsAuthorOrReadOnly]
+
+    def create(self, request, *args, **kwargs):
```
Comment (qmuz9ave) [inline posts/views.py:106]: better to use serializer and make function short

File: posts/views.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+                post = Post.objects.filter(id = post_id).first(),
+                user = request.user,
+                answer = data.get("answer"),
+            )
+            answer_obj.save()
+            serializer = AnswerSerializer(answer_obj)
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+        except Exception as e:
+            print(e)
```
Comment (qmuz9ave) [inline posts/views.py:126]: handle exception

File: quickflow/settings.py [hunk 15]
```diff
@@ -88,20 +104,20 @@
...
-        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
-    },
-    {
-        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
-    },
-    {
-        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
-    },
-]
+# AUTH_PASSWORD_VALIDATORS = [
```
Comment (qmuz9ave) [inline quickflow/settings.py:107]: remove comments

File: quickflow/settings.py [hunk 16]
```diff
@@ -142,3 +158,19 @@
...
 EMAIL_HOST_PASSWORD = config("EMAIL_HOST_PASSWORD")
+
+
+# Rest Framework section
+REST_FRAMEWORK = {
+    'DEFAULT_AUTHENTICATION_CLASSES': [
+        'rest_framework.authentication.BasicAuthentication',
+        'rest_framework.authentication.SessionAuthentication'
+    ],
+    # 'DEFAULT_PERMISSION_CLASSES': [
```
Comment (qmuz9ave) [inline quickflow/settings.py:169]: remove comments

File: posts/serializers.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        comments =  AnswerComment.objects.filter(answer = answer)
+        serializer = AnswerCommentSerializer(comments, many = True)
+        return serializer.data
+        
+    def create(self, validated_data):
+        post = validated_data.get('post')
+        answers = Answer.objects.filter(post = post)
+        for answer in answers:
+             if self.context['request'].user.id == answer.user.id:
+                raise serializers.ValidationError({'respone':'Answer was already given.'})
```
Comment (qmuz9ave) [inline posts/serializers.py:113]: detail

File: posts/views.py (hunk 14 above)
Comment (jiwan-gharti) [reply posts/views.py]: still incomplete so

**Review by qmuz9ave — APPROVED** [review]

**Pull request merged**